### ISSUE TITLE
SDPA decode produces incorrect results with batch-broadcast attention mask

### ISSUE THREAD

**Issue by jserbedzijaTT** [tenstorrent/tt-metal#39910]
## Summary

`ttnn.transformer.scaled_dot_product_attention_decode` produces incorrect results when the attention mask has a smaller batch dimension than the K/V tensors (e.g., mask `batch=1` broadcast across `batch=32` K/V). This affects non-causal cross-attention use cases such as Llama multimodal vision-language models.

## Root Cause

Two independent bugs were identified:

### Bug 1: Kernel mask batch offset uses K's batch size instead of mask's

In `reader_decode_all.cpp`, the mask DRAM tile offset is computed as:

```cpp
const uint32_t mask_batch_offset = ((cur_batch / q_heads_parallel_factor) % Bkv) * PNHt * St;
```

`Bkv` is K's batch size (e.g., 32), but the mask may have a different batch size (e.g., 1). When `mask_batch=1` and `Bkv=32`, batches 1–31 compute non-zero offsets (`1*PNHt*St`, `2*PNHt*St`, ...) that read past the end of the mask's DRAM buffer, producing garbage data.

### Bug 2: Program cache hash doesn't include the mask tensor

In `sdpa_decode_device_operation.cpp`, `compute_program_hash` hashes only a boolean `has_attn_mask` — not the mask tensor itself. This causes a program compiled for one mask shape/dtype to be incorrectly reused for a different mask. For example, a program compiled with a `[32, 1, 32, 128]` bf4b mask gets reused for a `[1, 1, 32, 128]` bf16 mask, producing wrong results for both shape and dtype reasons.

## Investigation

We verified the bugs are independent by testing each fix in isolation:

| Fix applied | bf4b full-batch | bf4b broadcast | bf16 full-batch | bf16 broadcast |
|---|---|---|---|---|
| Neither | PASS (0.9996) | FAIL (0.12) | FAIL (0.73) | FAIL (0.18) |
| Hash only | PASS (0.9996) | FAIL (0.13) | PASS (0.9996) | FAIL (0.75) |
| Both | PASS (0.9996) | PASS (0.9996) | PASS (0.9996) | PASS (0.9996) |

- The **hash fix** resolved bf16 full-batch failures (were cache collisions with bf4b programs)
- The **Bmask kernel fix** is required for broadcast — without it, the kernel reads out-of-bounds mask memory

Note: a comment in `models/tt_transformers/tt/multimodal/llama_cross_attention.py` confirms the team was already aware of related broadcast issues:
```python
# WARNING: this broadcast is also broken, must broadcast on host
```

## Fix

Commit: [`22b8c2e7868`](https://github.com/tenstorrent/tt-metal/commit/22b8c2e7868)
Branch: `jserbedzija/sdpa_mask_batch_broadcast_fix`

1. **Added `Bmask` compile-time arg** to the reader kernel, derived from `attn_mask->padded_shape()[0]`, and use it instead of `Bkv` for the mask batch offset calculation.

2. **Hash the mask tensor** in `compute_program_hash` instead of just hashing a boolean, so different mask shapes/dtypes get separate cached programs.

All 9 existing `test_sdpa_decode.py` tests continue to pass.

## Repro

A zero mask (`mask=0`) should produce identical results to no mask at all. On unfixed `main`, this gives PCC=0.068 instead of ~1.0:

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)
batch, nh, kv_seq, d = 32, 32, 128, 64

Q = torch.randn(1, batch, nh, d, dtype=torch.bfloat16)
K = torch.randn(batch, nh, kv_seq, d, dtype=torch.bfloat16)
V = torch.randn(batch, nh, kv_seq, d, dtype=torch.bfloat16)
mask = torch.zeros(1, 1, nh, kv_seq, dtype=torch.bfloat16)  # batch=1, broadcast

tt_Q = ttnn.from_torch(Q, layout=ttnn.TILE_LAYOUT, device=device)
tt_K = ttnn.from_torch(K, layout=ttnn.TILE_LAYOUT, device=device)
tt_V = ttnn.from_torch(V, layout=ttnn.TILE_LAYOUT, device=device)
tt_mask = ttnn.from_torch(mask, dtype=ttnn.bfloat4_b, layout=ttnn.TILE_LAYOUT, device=device)

output = ttnn.transformer.scaled_dot_product_attention_decode(
    tt_Q, tt_K, tt_V, is_causal=False, attn_mask=tt_mask, scale=d**-0.5)
actual = ttnn.to_torch(output)

# Compare against no-mask (zero mask should give same result)
output_nomask = ttnn.transformer.scaled_dot_product_attention_decode(
    tt_Q, tt_K, tt_V, is_causal=False, scale=d**-0.5)
expected = ttnn.to_torch(output_nomask)

pcc = torch.corrcoef(torch.stack([actual.float().flatten(), expected.float().flatten()]))[0, 1].item()
print(f"PCC (zero-mask vs no-mask): {pcc:.6f}")
assert pcc >= 0.99, f"FAIL: zero broadcast mask gives wrong results (PCC={pcc:.6f})"

ttnn.close_device(device)
```

**Comment by sdjordjevicTT** [comment]
@pavlejosipovic is this something that can be picked up any time soon?

**Comment by pavlejosipovic** [comment]
Second issue is solved on main, I'll take a look at the first one
